File: DataStructure/BinaryTree/Medium/113.PathSumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector <vector<int>> paths;
    int target;
    
    void findTargetPath(TreeNode* root, int currSum, vector<int> path) {
        if(NULL == root) {
            return;
        }

        currSum = currSum + root -> val;
        path.push_back(root -> val);

        if(NULL == root -> left && NULL == root -> right) {
            if(currSum == target) {
                paths.push_back(path);
            }

            return;
        }

        if (NULL != root -> left) {
            findTargetPath(root -> left, currSum, path);
        }

        if (NULL != root -> right) {
            findTargetPath(root -> right, currSum, path);
        }
    }
    
    
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if (NULL == root) {
            return paths;
        }

        target = targetSum;

        findTargetPath(root, 0, vector <int> ());

        return paths;
    }
};

File: DataStructure/BinaryTree/Medium/113.PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
	int target;
	List <List<Integer>> paths = new ArrayList<List<Integer>>();

	private void findTargetPath(TreeNode root, int currentSum, ArrayList<Integer> path) {
		if (root == null) {
			return;
		}

		currentSum = currentSum + root.val;
		path.add(root.val);

		if (root.left == null && root.right == null) {
			if (currentSum == target) {
				paths.add(new ArrayList<>(path));
			}

			// removing cause java always handles object as reference type
			path.remove(path.size() - 1); // backtracking
			return;
		}

		findTargetPath(root.left, currentSum, path);
		findTargetPath(root.right, currentSum, path);

		// removing cause java always handles object as reference type
		path.remove(path.size() - 1); // backtracking
	}

	public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
		if (null == root) {
			return paths;
		}

		target = targetSum;
		findTargetPath(root, 0, new ArrayList<Integer>());

		return paths;
	}
}